### DISCUSSION TITLE
Stuck server?

### DISCUSSION THREAD

**Discussion by rdica** [jamulussoftware/jamulus#2045] [Technical Support & Help]
There seems to be a stuck server in Genre Rock called `Thomakaju` at 85.214.171.92. It's been in this state for about a week. Is this something that can be removed from the Directory Servers?

**Comment by gilgongo** [comment]
That server is currently full (14/14) so you won't be able to join. The Jamulus client doesn't have a clear notification of this - is that what you mean? (see https://explorer.jamulus.io/)

**Reply by rdica** [reply]
No.
I don't believe the server is even up. I seriously doubt those same players have been constantly connected for a week. I believe the entry in the Directory is out of date and for some reason not removed.
An nmap scan of the server IP at udp port 22124 as reported by the explorer shows the port in a filtered state. This indicates some kind of firewalling.
Wouldn't that impact the ability of the Directory server to correctly index the server?

**Reply by gilgongo** [reply]
Ah I see. Sometimes people do hang out on servers for days on end but not that many at a time. Not sure if it's firewall related, the directory server in question is slighty different to the others I think so we'll investigate.

**Reply by softins** [reply]
The server `Thomakaju` must be up, because the Directory server does not store a list of clients connected to each registered server. So the Explorer back-end is getting the list of servers from the Directory, and is then querying `Thomakaju` for its client list, which is duly being returned.

So the question is: why are those client connections not timing out on `Thomakaju` and disappearing, if the clients are indeed no longer there?

**Reply by pljones** [reply]
There's definitely something weird with it.  I kicked it from the directory (with a little help from iptables) but it re-registered - it's definitely not totally dead.  Why it's still reporting all those clients, though, I don't know.  The network thread runs separately from the audio processing thread, and I think it's the latter that keeps the list updated.  So if that dies, I guess this could happen?

Ideally we'd get some diagnostics on it.  At a minimum, whoever runs the server needs to restart it.

**Reply by ghost** [reply]
There is also the possibility that the server administrator wants the server to operate this way. There are likely many loopholes in Jamulus protocol that programmers and server administrators can use for extra configuration options.

**Reply by rdica** [reply]
Interesting, thanks for the followup. I have nothing more to contribute to this, as far as I'm concerned this is anwswered.

**Reply by softins** [reply]
It would be very significant to know whether that server was using multi-threading (`-T`) or not. But without knowing whose server it is, it's hard to either find that out or to ask for the server to be restarted. Since the server is full, it's not even possible to fetch its welcome message, if any.

Looking at the code, if the clients are not timing out, that indicates `CServer::DecodeReceiveData()` is no longer being called. Without multi-threading enabled, that is called directly from `CServer::OnTimer()`. But with multi-threading enabled, it is called from the multiple threads in parallel.

The server in question is running 3.7.0. The multi-threading code was changed significantly between 3.7.0 and 3.8.0, so it may well be that 3.8.0 would not suffer from the same problem.

**Reply by pljones** [reply]
> that is called directly from `CServer::OnTimer()`.

Could the Qt timer die?  Or become disconnected?  I'd have expected to see problems with other servers, though, given how many there are.

Whilst the server seems to be running vanilla Jamulus 3.7.0 code, there's no telling whether it was installed from the supplied binaries or locally compiled, which would open up even further possibilities.
